src/expr/impl/src/scalar/to_jsonb.rs

```diff
@@ -0,0 +1,295 @@
+    input: StructRef<'_>,
+    data_type: &DataType,
+    builder: &mut Builder,
+) -> Result<()> {
+    builder.begin_object();
+    let names: Vec<&str> = data_type.as_struct().names().collect();
+    input
+        .iter_fields_ref()
+        .zip_eq_fast(data_type.as_struct().types())
+        .enumerate()
+        .map(|(idx, (scalar, data_type))| {
+            if names.is_empty() {
+                builder.display(format_args!("f{}", idx + 1));
+            } else {
+                builder.add_string(names[idx]);
+            };
+            add_scalar_to_builder(scalar, data_type, builder)?;
+            Ok::<(), ExprError>(())
+        })
+        .try_collect()?;
+    builder.end_object();
+    Ok(())
+}
+
+fn add_scalar_to_builder(
+    input: Option<ScalarRefImpl<'_>>,
+    data_type: &DataType,
+    builder: &mut Builder,
+) -> Result<()> {
+    match input {
+        Some(input) => match input {
+            ScalarRefImpl::Int16(v) => v.add_to(builder)?,
+            ScalarRefImpl::Int32(v) => v.add_to(builder)?,
+            ScalarRefImpl::Int64(v) => v.add_to(builder)?,
+            ScalarRefImpl::Int256(v) => v.add_to(builder)?,
+            ScalarRefImpl::Float32(v) => v.add_to(builder)?,
+            ScalarRefImpl::Float64(v) => v.add_to(builder)?,
+            ScalarRefImpl::Utf8(v) => v.add_to(builder)?,
+            ScalarRefImpl::Bool(v) => v.add_to(builder)?,
+            ScalarRefImpl::Decimal(v) => v.add_to(builder)?,
+            ScalarRefImpl::Interval(v) => v.add_to(builder)?,
+            ScalarRefImpl::Date(v) => v.add_to(builder)?,
+            ScalarRefImpl::Time(v) => v.add_to(builder)?,
+            ScalarRefImpl::Timestamp(v) => v.add_to(builder)?,
+            ScalarRefImpl::Jsonb(v) => v.add_to(builder)?,
+            ScalarRefImpl::Serial(v) => v.add_to(builder)?,
+            ScalarRefImpl::Bytea(v) => v.add_to(builder)?,
+            ScalarRefImpl::Timestamptz(v) => v.add_to(builder)?,
+            ScalarRefImpl::Struct(v) => add_struct_to_builder(v, data_type, builder)?,
+            ScalarRefImpl::List(v) => add_anyarrary_to_builder(v, data_type, builder)?,
+        },
+        None => builder.add_null(),
+    };
+    Ok(())
+}
+
+/// Values that can be converted to JSONB.
+pub trait ToJsonb {
+    fn add_to(self, builder: &mut Builder) -> Result<()>;
+}
+
+impl ToJsonb for bool {
+    fn add_to(self, builder: &mut Builder) -> Result<()> {
+        builder.add_bool(self);
+        Ok(())
+    }
+}
+
+impl ToJsonb for i16 {
+    fn add_to(self, builder: &mut Builder) -> Result<()> {
+        builder.add_i64(self as _);
+        Ok(())
+    }
+}
+
+impl ToJsonb for i32 {
+    fn add_to(self, builder: &mut Builder) -> Result<()> {
+        builder.add_i64(self as _);
+        Ok(())
+    }
+}
+
+impl ToJsonb for i64 {
+    fn add_to(self, builder: &mut Builder) -> Result<()> {
+        let res: F64 = self
+            .try_into()
+            .map_err(|_| ExprError::CastOutOfRange("IEEE 754 double"))?;
```

Comment:
Just sharing more on this, given the e2e test failure in CI:
* RFC 8259 allows this: `type Number = f64`.
* PostgreSQL does this: `type Number = Decimal`.
  * PostgreSQL decimal can even hold `f64::MAX`
* serde_json/simd-json/jsonbb does this: `enum Number {I64, U64, F64}`
  * serde_json has a feature `arbitrary_precision` to `type Number = String`
  * simd-json has a feature `128bit` to `enum Number {I64, U64, F64, I128, U128}`
  * jsonbb may be extended to `enum Number {I64, U64, F64, Decimal}`
    * This requires a variable length decimal. `rust_decimal` is not enough.

Back to `to_jsonb`: we do have the capability to avoid precision loss on i64 today, but using the lossy f64 feels more consistent with decimal / int256 which we cannot encode losslessly yet. Both sounds acceptable to me. And the `.0` issue would cause less problems after the jsonbb serializer fix.